L1Trigger/TrackerTFP/src/SeedFilter.cc

```diff
@@ -0,0 +1,242 @@
+      TTDTC::Stream& lostStream = lost[region_ * dataFormats_->numChannel(Process::mht) + channel];
+      deque<StubSF*> output;
+      vector<StubMHT*>& stubs = input_[channel];
+      stubs.erase(remove(stubs.begin(), stubs.end(), nullptr), stubs.end());
+      set<int> ids;
+      for (StubMHT* stub : stubs)
+        if (stub)
+          ids.insert(stub->trackId());
+      vector<vector<StubMHT*>> tracks(ids.size());
+      int i(0);
+      for (auto it = stubs.begin(); it != stubs.end();) {
+        const auto start = it;
+        const int id = (*it)->trackId();
+        auto different = [id](StubMHT* stub){ return id != stub->trackId(); };
+        it = find_if(it, stubs.end(), different);
+        vector<StubMHT*>& track = tracks[i++];
+        track.reserve(distance(start, it));
+        copy(start, it, back_inserter(track));
+      }
+      for (const vector<StubMHT*>& track : tracks) {
+        const int eta = track.front()->sectorEta();
+        const double dZT = (sinh(setup_->boundarieEta(eta + 1)) - sinh(setup_->boundarieEta(eta))) * setup_->chosenRofZ();
+        vector<vector<StubMHT*>> layerStubs(numLayers);
+        for (vector<StubMHT*>& layer : layerStubs)
+          layer.reserve(track.size());
+        for (StubMHT* stub : track)
+          layerStubs[layerId(stub)].push_back(stub);
+        deque<pair<StubMHT*, StubMHT*>> seeds;
+        for (const pair<int, int>& seedingLayer : seedingLayers)
+          for (StubMHT* inner : layerStubs[seedingLayer.first])
+            if (setup_->psModule(inner->ttStubRef()))
+              for (StubMHT* outer : layerStubs[seedingLayer.second])
+                if (setup_->psModule(outer->ttStubRef()))
+                  if (outer->r() > inner->r())
+                    seeds.emplace_back(inner, outer);
+        vector<vector<StubSF*>> seedTracks;
+        seedTracks.reserve(seeds.size());
+        for (const pair<StubMHT*, StubMHT*>& seed : seeds) {
+          const double r1 = seed.first->r() + dT;
+          const double r2 = seed.second->r() + dT;
+          const double z1 = seed.first->z();
+          const double z2 = seed.second->z();
+          const double r = (r2 + r1) / 2.;
+          const double z = (z2 + z1) / 2.;
+          const double dr = (r2 - r1);
+          const double dz = (z2 - z1);
+          const double cot = cot_.digi(dz / dr);
+          const double cotGlobal = cot + setup_->sectorCot(eta);
+          const double zT = zT_.digi(z - cot * r);
+          const double z0 = zT - cot * setup_->chosenRofZ();
+          if (abs(z0) > setup_->beamWindowZ() + cot_.base() * setup_->chosenRofZ())
+            continue;
+          if (abs(zT) > dZT / 2. + zT_.base() / 2.)
+            continue;
+          //if (!cot_.inRange(cot))
+            //continue;
+          const vector<int>& layerEncoding = layerEncoding_->layerEncoding(eta, zT_.toUnsigned(zT_.integer(zT)), cot_.toUnsigned(cot_.integer(cot)));
```

Comment:
Add comment "Ordered list of which tracker layers (barrel=1-6, endcap=11-15) particle of this (zT,cot )is expected to cross".